File: WotjaAPI_WinRT_demo/WotjaAPI.cpp
//
// Wotja_API.cpp
// WotjaAPI_WinRT_demo
//
// Copyright (C) 2007-2024 Intermorphic Limited. All rights reserved.
// Licensed under the MIT License.
// Note: For the avoidance of doubt use of Wotja is governed by the Wotja EULA to be found at https://wotja.com/legal/eula
//

#include "pch.h"

#include "WotjaAPI.h"

#include <locale>
#include <codecvt>
#include <string>
#include <fstream>
#include <iostream>

#include <appmodel.h>

using namespace winrt;
using namespace Windows::Foundation;

static bool mbPlaying = false;

static void sOpenURL(const std::wstring& urlString) {
  winrt::Windows::Foundation::Uri uri(urlString);
  winrt::Windows::System::Launcher::LaunchUriAsync(uri);
}

static void sOpenURLInBackground(const std::wstring& urlString) {
  // There is no specific way to do this on Windows.
  // What happens, depends on the target URL handler.
  // In general, Wotja launches in the background.
  sOpenURL(urlString);
}

static void sOpenURLInForeground(const std::wstring& urlString) {
  // There is no specific way to do this on Windows.
  // What happens, depends on the target URL handler.
  // In general, Edge etc. launch in the background.
  sOpenURL(urlString);
}

static std::wstring sTemporaryFilePath = L"";
void WotjaAPI::setFilePathForURLSend(const std::wstring& temporaryFilePath) {
  sTemporaryFilePath = temporaryFilePath;
}

#include <algorithm>

void replaceAll(std::wstring& str, const std::wstring& from, const std::wstring& to) {
  size_t start_pos = 0;
  while ((start_pos = str.find(from, start_pos)) != std::wstring::npos) {
    str.replace(start_pos, from.length(), to);
    start_pos += to.length();
  }
}

std::wstring convertNarrowToWide(const std::string& sString) {
  int sLength = (int)sString.length() + 1;
  int wLength = MultiByteToWideChar(CP_ACP, 0, sString.c_str(), sLength, 0, 0);
  std::wstring wString(wLength, L'\0');
  MultiByteToWideChar(CP_ACP, 0, sString.c_str(), sLength, &wString[0], wLength);
  return wString;
}

std::string convertWideToNarrow(const std::wstring& wString) {
  auto numBytes = WideCharToMultiByte(CP_UTF8, 0, wString.c_str(), -1, nullptr, 0, nullptr, nullptr);

  std::string nString(numBytes + 1, '\0');
  WideCharToMultiByte(CP_UTF8, 0, wString.c_str(), -1, &nString[0], numBytes, nullptr, nullptr);
  return nString;
}

std::string encodeStringForWotjaURI(const std::string& sString) {
  // The only valid characters for a Wotja URI are those in the following character set:
  auto cWotjaURI_ValidCharsSet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
  std::string validCharSet(cWotjaURI_ValidCharsSet);

  std::stringstream encodedPathStream;
  for (auto theChar : sString) {
    if (validCharSet.find(theChar) != std::string::npos) {
      encodedPathStream << theChar;
    } else if (theChar == '\0') {
    } else {
      encodedPathStream << "%";
      encodedPathStream << std::uppercase << std::setfill('0') << std::setw(2) << std::hex << (int)theChar;
    }
  }
  auto result = encodedPathStream.str();
  return result;
}

static void loadAndPlayFromFile(const std::wstring& wotjaURIString) {
  std::string pathStr  = convertWideToNarrow(sTemporaryFilePath);
  std::string wotjaStr = convertWideToNarrow(wotjaURIString);

  // Replace invalid path characters, with encoded characters.
  std::string sPathToWrite = encodeStringForWotjaURI(pathStr);
  auto wPathToWrite = convertNarrowToWide(sPathToWrite);

  std::wstringstream stringstream;
  stringstream << L"wotja://api_load/?path=" << wPathToWrite;
  std::wstring urlToSend = stringstream.str();

  std::ofstream outStream(sTemporaryFilePath, std::ios::binary | std::ios::trunc);
  outStream << wotjaStr;
  outStream.close();

  mbPlaying = true;

  sOpenURLInBackground(urlToSend);
}

void WotjaAPI::loadAndPlay(const std::wstring& wotjaURIString) {

  // Load Wotja content, and start it playing.
  if (sTemporaryFilePath != L"") {
    // Use the path we've been provided
    loadAndPlayFromFile(wotjaURIString);
    return;
  }

  std::wstring useString = wotjaURIString;
  if (useString.starts_with(L"wotja://api_load/?data=") == false) {
    // Change wotja://blahblah to wotja://api_load/?data=blahblah
    useString.replace(0, 8, L"wotja://api_load/?data=");
  }

  mbPlaying = true;

  sOpenURLInBackground(useString);
}

void WotjaAPI::stop() {
  mbPlaying = false;
  sOpenURLInBackground(L"wotja://api_stop/");
}

void WotjaAPI::playWithTrackIndex(int trackIndex) {
  mbPlaying = true;

  std::wstringstream stringstream;
  stringstream << L"wotja://api_play/?ti=" << trackIndex;
  sOpenURLInBackground(stringstream.str());
}

bool WotjaAPI::getIsPlaying() {
  return mbPlaying;
}

// MARK: Wotja availability checking

// Wotja API support added in the following Wotja version
static std::wstring cMinimumWotjaVersion = L"24.0.90";

static WotjaAPI::WotjaStatus sWotjaStatus        = WotjaAPI::WotjaStatus_looking_for_wotja;
static std::wstring sWotjaVersion                = L""; // Set internally to e.g. "24.1.0"
static std::wstring sWotjaMinimumVersionRequired = L""; // Set by app to e.g. "24.1.0"

// Returns e.g. "Looking for Wotja..."
std::wstring WotjaAPI::convertWotjaStatusToString(WotjaAPI::WotjaStatus wotjaStatus) {
  switch (wotjaStatus) {
  case WotjaAPI::WotjaStatus_not_found:
    return L"Wotja not found";
  case WotjaAPI::WotjaStatus_needs_update:
    return std::wstring(L"Wotja update required (found ") + sWotjaVersion + L"), needs " + sWotjaMinimumVersionRequired;
  case WotjaAPI::WotjaStatus_found:
    return std::wstring(L"Wotja found (version ") + sWotjaVersion + L")";
  case WotjaAPI::WotjaStatus_looking_for_wotja:
  default:
    return L"Looking for Wotja...";
  }
}

// Returns e.g. @"Looking for Wotja..."
std::wstring WotjaAPI::getWotjaStatusString() {
  return convertWotjaStatusToString(getWotjaStatus());
}

// Returns e.g. @"24.1.0"
std::wstring WotjaAPI::getFirstSupportedWotjaVersion() {
  return cMinimumWotjaVersion;
}

WotjaAPI::WotjaStatus WotjaAPI::getWotjaStatus() {
  return sWotjaStatus;
}

uint32_t versionNumberFromString(const std::wstring& value) {
  std::vector<std::wstring> output;

  std::wstring::size_type prev_pos = 0, pos = 0;

  while ((pos = value.find(L".", pos)) != std::string::npos) {
    std::wstring substring(value.substr(prev_pos, pos - prev_pos));
    output.push_back(substring);
    prev_pos = ++pos;
  }

  output.push_back(value.substr(prev_pos, pos - prev_pos)); // Last word

  uint32_t theNumber = 0;

  for (auto& component : output) {
    uint32_t testChar = (uint32_t)std::stoi(component);
    theNumber         = (theNumber << 8) + testChar;
  }

  return theNumber;
}

void WotjaAPI::searchForWotja(const std::wstring& withMinimumVersion, std::function<void(WotjaStatus)> completion) {
  // This works, from the Power Shell!
  // Get-AppxPackage -Name "Intermorphic.Wotja22Lite"

  sWotjaMinimumVersionRequired = withMinimumVersion;

  std::thread  t1([completion] {
    auto   packageFamilyName = L"Intermorphic.Wotja22Lite_2dswjjh8nep0p";
    UINT32 count = 0, bufferLength = 0;

    GetPackagesByPackageFamily(packageFamilyName, &count, 0, &bufferLength, 0);
    if (count > 0) {
      PWSTR* packageFullNames = (PWSTR*)alloca(count * sizeof(PWSTR) + bufferLength * sizeof(WCHAR));
      PWSTR  buffer           = (PWSTR)(packageFullNames + count);

      if (NOERROR == GetPackagesByPackageFamily(packageFamilyName, &count, packageFullNames, &bufferLength, buffer)) {
        if (count > 0) {
          PCWSTR packageFullName = *packageFullNames++;
          OutputDebugStringW(L"packageFullName=");
          OutputDebugStringW(packageFullName);
          OutputDebugStringW(L"\n");

          std::wstring stripPrefix(L"Intermorphic.Wotja22Lite_");
          std::wstring versionString = packageFullName;
          if (versionString.starts_with(stripPrefix)) {
            versionString.replace(0, stripPrefix.length(), L"");
          }

          std::wstring stripSuffix(L"_x64__2dswjjh8nep0p");
          if (versionString.ends_with(stripSuffix)) {
            versionString.erase(versionString.length() - stripSuffix.length());
          }

          std::wstring stripSuffix2(L".0");
          if (versionString.ends_with(stripSuffix2)) {
            versionString.erase(versionString.length() - stripSuffix2.length());
          }

          sWotjaVersion = versionString;

          auto versionNumber = versionNumberFromString(sWotjaVersion);

          uint32_t minimumRequiredVersionNumber = versionNumberFromString(sWotjaMinimumVersionRequired);
          if (versionNumber < minimumRequiredVersionNumber) {
            sWotjaStatus = WotjaStatus_needs_update;
            completion(WotjaStatus_needs_update);
          } else {
            sWotjaStatus = WotjaStatus_found;
            completion(WotjaStatus_found);
          }

          return;
        }
      }

      sWotjaStatus = WotjaStatus_not_found;
      completion(sWotjaStatus);
    }
  });
  t1.join();
}

// MARK: Wotja installation, update and miscellaneous

void WotjaAPI::installWotja() {
  sOpenURLInForeground(L"https://apps.microsoft.com/store/detail/9NH2TJ3JGNVP");
}

void WotjaAPI::updateWotja() {
  sOpenURLInForeground(L"https://apps.microsoft.com/store/detail/9NH2TJ3JGNVP");
}

void WotjaAPI::showWotjaEULA() {
  sOpenURLInForeground(L"https://wotja.com/legal/eula");
}

void WotjaAPI::showDeveloperFAQ() {
  sOpenURLInForeground(L"https://wotja.com/help/#developer");
}

void WotjaAPI::showAboutWotja() {
  sOpenURLInForeground(L"https://wotja.com/music/");
}

void WotjaAPI::showWotja() {
  sOpenURLInForeground(L"wotja://api_show/");
}

File: WotjaAPI_SwiftUI_demo/WotjaAPI.m
//
// WotjaAPI.m
// WotjaAPI_SwiftUI_demo
//
// Copyright © 1990-2026 Intermorphic Limited. All Rights Reserved.
// Licensed under the MIT License.
// Note: For the avoidance of doubt use of Wotja is governed by the Wotja EULA to be found at https://wotja.com/legal/eula
//

#include "WotjaAPI.h"

#import <CoreAudioKit/CoreAudioKit.h>
#include <AudioUnit/AudioUnit.h>

@interface WotjaAPI()
@end

#if defined(IM_TARGET_IOS)
#define USE_MIDI_PORT 1
#endif // defined(IM_TARGET_IOS)

#if defined(IM_TARGET_MACOS)
// No need to use MIDI port on macOS, as NSWorkspace openURL works fine.
//#define USE_MIDI_PORT 1
#endif // defined(IM_TARGET_MACOS)

@implementation WotjaAPI

- (instancetype)init {
  self = [super init];
  if (self) {
  }
  return self;
}

// Private static method.
// Make an openURL call for iOS or macOS, in a platform specific way.
+(void) sOpenURL:(NSString*)urlString activates:(BOOL)activates {
  NSURL* url = [NSURL URLWithString:urlString];

#if defined(IM_TARGET_IOS)
  // On iOS, there is currently no way to prevent giving Wotja the UI focus - activates is ignored.
  // The user must manually "go back" to the calling application after any call to openURL.
  [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:^(BOOL success) {
  }];
#elif defined(IM_TARGET_MACOS)
  NSWorkspaceOpenConfiguration* configuration = [NSWorkspaceOpenConfiguration configuration];
  configuration.activates = activates;

  [[NSWorkspace sharedWorkspace] openURL:url configuration:configuration completionHandler:^(NSRunningApplication * _Nullable app, NSError * _Nullable error) {
  }];
#endif // defined(TARGET_OS_MAC)
}

+(void) sOpenURLInForeground:(NSString*)urlString {
  [self sOpenURL:urlString activates:YES];
}

+(void) sOpenURLInBackground:(NSString*)urlString {
  // On iOS, there is currently no way to open Wotja in the background.
  // The user must manually "go back" to the calling application after any call to openURL.
  // On macOS, this works as expected.
  [self sOpenURL:urlString activates:NO];
}

// MARK: MIDI port handling (internal code)

#if defined(USE_MIDI_PORT)
// MIDI client, port and endpoint
static MIDIClientRef midiClient = 0;
static MIDIPortRef midiOutputPort = 0;
static MIDIEndpointRef wotjaEndpointRef = 0;
// State: record if we've sent the first command via a URL, or not.
static BOOL bSentFirstURL = false;

// Structure and callback; used to tidy-up the MIDI message data after it has been sent

typedef struct MySysexRequest_t {
  void* mData;
} MySysexRequest;

static void completionProc(MIDISysexSendRequest *request)
{
  if(request->complete) {
    MySysexRequest* mySysexRequest = (MySysexRequest*)request->completionRefCon;

    free(mySysexRequest->mData);
    free((void*)request);
  }
}

+(NSString*) sGetMidiObjectDisplayName:(MIDIObjectRef)endpoint {
  CFStringRef result = CFSTR("");

  MIDIObjectGetStringProperty(endpoint, kMIDIPropertyDisplayName, &result);

  if (result == nil) {
#if DEBUG
    NSLog(@"DEBUG: warning - kMIDIPropertyDisplayName returned nil");
#endif // DEBUG
    return @"Unknown";
  }

  NSString *resultNSString = (__bridge NSString *)result;

  CFRelease(result);

#if DEBUG
  //NSLog(@"DEBUG: kMIDIPropertyDisplayName is (%@)", resultNSString);
#endif // DEBUG

  return resultNSString;
}

void clientCompletionProc(const struct MIDINotification * notif , void * other ) {
  //NSLog(@"clientCompletionProc");
}

+(MIDIEndpointRef) sGetWotjaInputMIDIEndpointRef {
  unsigned long count = MIDIGetNumberOfDestinations();
  for (unsigned long i = 0; i < count; i++) {
    MIDIEndpointRef endpoint = MIDIGetDestination(i);
    if (endpoint != 0) {
      NSString*  name = [WotjaAPI sGetMidiObjectDisplayName:endpoint];
      if ([name isEqualToString:@"Wotja Input"]) {
        return endpoint;
      }
    }
  }

  return 0;
}

/// A helper that NSLogs an error message if "c" is an error code
void NSLogError(OSStatus c, NSString *str)
{
  if (c) {
    NSLog(@"Error (%@): %ld:%@", str, (long)c,[NSError errorWithDomain:NSMachErrorDomain code:c userInfo:nil]);
  }
}
#endif // defined(USE_MIDI_PORT)

// MARK: Public API method - make a Wotja API call.

// Note: the first call is always made via openURL, which on iOS will always give the Wotja App the UI focus; the user will have to "go back" the the calling app.
// Subsequent calls are made (where possible) via Wotja's "Wotja Input" virtual MIDI port, which won't give Wotja the UI focus (even on iOS).

+(BOOL) sMakeWotjaAPICall:(NSString*)wotjaURIString {

  if ([wotjaURIString hasPrefix:@"wotja://api_"] == false) {
    return NO;
  }

#ifdef DEBUG
  // NSLog(@"DEBUG: sMakeWotjaAPICall: processing Wotja URI of size %d characters, UTF8 string length=%d", (int)wotjaURIString.length, (int)[wotjaURIString lengthOfBytesUsingEncoding:NSUTF8StringEncoding]);
#endif // DEBUG

#if !defined(USE_MIDI_PORT)
  [WotjaAPI sOpenURLInBackground:wotjaURIString];
#elif defined(USE_MIDI_PORT)
  if (wotjaEndpointRef == 0) {
    wotjaEndpointRef = [WotjaAPI sGetWotjaInputMIDIEndpointRef];
  }

  if (wotjaEndpointRef != 0 && midiClient == 0) {
    // clientCompletionProc is called only if there is a NSRunLoop for this thread...
    OSStatus status = MIDIClientCreate(CFSTR("My Client"), clientCompletionProc, NULL, &midiClient);
    if (status != noErr) {
#if DEBUG
      NSLog(@"DEBUG: MIDIClientCreate failed!");
#endif // DEBUG
    } else {
      //status = MIDIOutputPortCreate(midiClient, CFSTR("My Port"), &midiOutputPort);
      status = MIDIInputPortCreateWithProtocol(midiClient, CFSTR("My Port"), kMIDIProtocol_2_0, &midiOutputPort, ^(const MIDIEventList * _Nonnull evtlist, void * _Nullable srcConnRefCon) {
#if DEBUG
          NSLog(@"DEBUG: MIDIInputPortCreateWithProtocol completed");
#endif // DEBUG
      });
      if (status != noErr) {
#if DEBUG
        NSLog(@"DEBUG; MIDIInputPortCreateWithProtocol failed!");
#endif // DEBUG
      }
    }
  }

  //
  // The Intial Wotja API call is made via the system openURL mechanism.
  // The call to openURL is also used if we were unable to establish a MIDI connection.
  //
  if (bSentFirstURL == false || midiClient == 0 || midiOutputPort == 0 || wotjaEndpointRef == 0) {
    // No MIDI; or very first URL; communicate using openURL.
    if (bSentFirstURL == false) {
#if DEBUG
      NSLog(@"DEBUG: Sending first wotja://api... as URL");
#endif // DEBUG
      bSentFirstURL = true;
    } else {
#if DEBUG
      NSLog(@"DEBUG: Warning: could not find Wotja Input MIDI port - sending second or later wotja://api... as URL");
#endif // DEBUG
    }
    [WotjaAPI sOpenURLInBackground:wotjaURIString];
    return NO;
  }

  //
  // Subsequent Wotja API calls are made via the Wotja Input MIDI port
  // (assuming we could find it!)
  //

  NSMutableData *sysexData = [NSMutableData new];
  char bytesToAppend[] = {
    0xf0, // Sysex start
    0x7D  // Manufacturer byte 1 (only)
  };
  [sysexData appendBytes:bytesToAppend length:sizeof(bytesToAppend)];

  NSData* apiData = [wotjaURIString dataUsingEncoding:NSUTF8StringEncoding];
  [sysexData appendData:apiData];

  char bytesToTerminate[] = {
    0xf7 // Sysex end
  };
  [sysexData appendBytes:bytesToTerminate length:sizeof(bytesToTerminate)];

  int bytesToSend = (int)[sysexData length];
  Byte *dataToSend = (Byte *)[sysexData bytes];

  // Use MIDISendSysex
  {
    Byte* theData = malloc([sysexData length]);
    memcpy(theData, dataToSend, bytesToSend);

    MySysexRequest* mySysexRequest = (MySysexRequest*)calloc(sizeof(MySysexRequest), 1);
    mySysexRequest->mData = theData;

    MIDISysexSendRequest* sendRequest = (MIDISysexSendRequest*)calloc(sizeof(MIDISysexSendRequest), 1);
    sendRequest->destination = wotjaEndpointRef;
    sendRequest->data = theData;
    sendRequest->bytesToSend = bytesToSend;
    sendRequest->complete = 0;
    sendRequest->completionProc = completionProc;
    sendRequest->completionRefCon = mySysexRequest;

    MIDISendSysex(sendRequest);
  }

  // Use MIDISend - messages sent like are never received by Wotja
  /*
  {
    int maxChunkSize = 512; // 65536;
    int packetSize = bytesToSend;
    int chunkSize = bytesToSend;
    if (packetSize <= maxChunkSize) {
      chunkSize = packetSize;
    } else {
      chunkSize = maxChunkSize;
    }

    int byteOffset = 0;
    int bytesRemaining = packetSize;
    while (bytesRemaining > 0) {
      // The maximum size is 65536 ... larger messages, must be split into separate packets!

      int bytesThisTime = bytesRemaining;
      if (bytesThisTime > chunkSize) {
        bytesThisTime = chunkSize;
      }

      {
        int PACKETLIST_SIZE = 1024 + sizeof(MIDIPacketList);
        MIDIPacketList* midiPacketList = (MIDIPacketList*)malloc(PACKETLIST_SIZE);
        //MIDIPacketList midiPacketList = {};
        MIDIPacket *currPacket = MIDIPacketListInit(midiPacketList);

        // If you want to send the MIDI data immediately, just provide a timestamp of 0.
        // If you call MIDIPacketListAdd with the same timestamp as the previous call, then it just concatenates your data to the previous packet. You will see packetList->numPackets and currentPacket stay the same, but currentPacket->length will increase.

        uint64_t timestamp = mach_absolute_time();

        currPacket = MIDIPacketListAdd(midiPacketList, PACKETLIST_SIZE, currPacket, timestamp, bytesThisTime, dataToSend + byteOffset);
        OSStatus status = MIDISend(midiOutputPort, wotjaEndpointRef, midiPacketList);
        if (status != noErr) {
          NSLogError(status, @"MIDISend");
        }

        // TODO - does this "free" prevent the MIDI being sent?
        //free(midiPacketList);

        bytesRemaining -= bytesThisTime;
        byteOffset += bytesThisTime;
      }
    }
  }
  */

  /*
  {
    MIDIEventList midiEventList = {};
    MIDIEventPacket* endPacket = MIDIEventListInit(&midiEventList, kMIDIProtocol_1_0);
    // The maximum size of a event list is 65536 bytes. Large sysex messages must be sent in smaller event lists.
    UInt32 *newMsg = [self convertMidi1ToMidi2WithData:pMsg size:size]; // convert byte array to UInt32 array
      event = MIDIEventListAdd(eventList, bufSize, event, 0, size, newMsg);
    endPacket = MIDIEventListAdd(&midiEventList,  65536, endPacket, 0, ump.count, ump)
    MIDISendEventList(outputPort, destination, &eventList)
  }
   */

  sysexData = nil;
#endif // !defined(USE_MIDI_PORT)

  return YES;
}

static bool mbPlaying = NO;

+(void) loadAndPlay:(NSString*)wotjaURIString trackIndex:(NSInteger)trackIndex tempo:(NSString*)tempo root:(NSString*)root {
  mbPlaying = YES;

  // Load and play using the provided Wotja URI
  if ([wotjaURIString hasPrefix:@"wotja://api_"] == YES) {
#ifdef DEBUG
    NSLog(@"Invalid string format!");
#endif // DEBUG
    return;
  }

  NSString* wotjaData = [wotjaURIString stringByReplacingOccurrencesOfString:@"wotja://" withString:@""];
  NSMutableString* apiURIString = [NSMutableString stringWithFormat:@"wotja://api_load/?data=%@", wotjaData];

  if (trackIndex < 0) {
    // Ignore it!
#ifdef DEBUG
    NSLog(@"Invalid track index!");
#endif // DEBUG
  } else {
    [apiURIString appendString: [NSString stringWithFormat:@"&ti=%d", (int)trackIndex]];
  }

  if (tempo == nil || [tempo isEqualToString:@"-"] || tempo.length == 0 || tempo.intValue < 0) {
    // Ignore it!
  } else {
    [apiURIString appendString: [NSString stringWithFormat:@"&tempo=%@", tempo]];
  }

  if (root == nil || [root isEqualToString:@"-"] || root.length == 0 ) {
    // Ignore it!
  } else {
    [apiURIString appendString: [NSString stringWithFormat:@"&root=%@", root]];
  }

#ifdef DEBUG
  //NSLog(@"loadAndPlay: apiURIString=%@", apiURIString);
#endif // DEBUG

  [self sMakeWotjaAPICall:apiURIString];
}

+(BOOL) getIsPlaying {
  return mbPlaying;
}

+(void) stop {
  mbPlaying = NO;

  [self sMakeWotjaAPICall:@"wotja://api_stop/"];
}

+(void) playWithTrackIndex:(NSInteger)trackIndex {
  mbPlaying = YES;

  [self sMakeWotjaAPICall:[NSString stringWithFormat:@"wotja://api_play/?ti=%d", (int)trackIndex]];
}

+(void) setTempo:(NSString*)tempo {
  if ([tempo isEqualToString:@"-"] || tempo.length == 0 || tempo.intValue < 0) {
    // Ignore it!
    return;
  }

  NSMutableString* apiURIString = [NSMutableString stringWithFormat:@"wotja://api_set/?tempo=%d", tempo.intValue];

  [self sMakeWotjaAPICall:apiURIString];
}

+(void) setRoot:(NSString*)root {
  if ([root isEqualToString:@"-"] || root.length == 0) {
    // Ignore it!
    return;
  }

  NSMutableString* apiURIString = [NSMutableString stringWithFormat:@"wotja://api_set/?root=%@", root];

  [self sMakeWotjaAPICall:apiURIString];
}


// MARK: Wotja availability checking

// Wotja API support added in the following Wotja version
static NSString* cMinimumWotjaVersion = @"26.0.0";

static WotjaStatus sWotjaStatus = WotjaStatus_looking_for_wotja;
static NSString* sWotjaVersion = @""; // Set internally to e.g. "26.0.0"
static NSString* sWotjaMinimumVersionRequired = @""; // Set by app to e.g. "26.0.0"

+(NSString*) getFirstSupportedWotjaVersion {
  return cMinimumWotjaVersion;
}

+(NSString*) getWotjaStatusString {
  return [self convertWotjaStatusToString:sWotjaStatus];
}

+(NSString*) convertWotjaStatusToString:(WotjaStatus)wotjaStatus {
  switch (wotjaStatus) {
  case WotjaStatus_not_found:
    return @"Wotja not found";
  case WotjaStatus_needs_update:
    return [NSString stringWithFormat:@"Wotja update required (found %@), needs %@", sWotjaVersion, sWotjaMinimumVersionRequired];
  case WotjaStatus_found:
    return [NSString stringWithFormat:@"Wotja found (version %@)", sWotjaVersion];
  case WotjaStatus_looking_for_wotja:
  default:
    return @"Looking for Wotja...";
  }
}

+(WotjaStatus) getWotjaStatus {
  return WotjaStatus_found; // sWotjaStatus;
}

+(void) searchForWotja:(NSString*)withMinimumVersion completion:(void(^_Nonnull)(WotjaStatus))completion {
  sWotjaMinimumVersionRequired = withMinimumVersion;

  dispatch_async(dispatch_get_main_queue(), ^{
    [self sPrivateSearchForWotja:withMinimumVersion completion:^(WotjaStatus result) {
      sWotjaStatus = result;
      completion(result);
    }];
  });
}

// MARK: Private methods

NSString* versionNumberToString(UInt32 versionNumber) {
  return [NSString stringWithFormat:@"%d.%d.%d", (int)(versionNumber>>16), (int)((versionNumber>>8) & 0xff), (int)(versionNumber & 0xff)];
}

UInt32 versionNumberFromString(NSString* value) {
  NSArray<NSString*>* components = [value componentsSeparatedByString:@"."];
  UInt32 theNumber = 0;

  for (NSString* component in components) {
    UInt32 testChar = (UInt32)[component integerValue];
    theNumber = (theNumber << 8) + testChar;
  }

  return theNumber;
}

+(void) sPrivateSearchForWotja:(NSString*)withMinimumVersion completion:(void(^_Nonnull)(WotjaStatus))completion {

  // On iOS, we found that canOpenURL doesn't work reliably for detecting the Wotja App.
  // However, scanning for the Wotja AUv3 works reliably (on both iOS and macOS).
  // The only way we found on iOS to check for Wotja version,
  // is to check for the AUv3 version; this also works well on macOS.
  // Therefore, we detect Wotja, and report the Wotja installed version,
  // by checking for the Wotja AUv3 and (if found) querying the Wotja AUv3 version.

  // Describe the Wotja AUv3 unit
  AudioComponentDescription audioComponentDescription;
  audioComponentDescription.componentType = 0; // Wildcard
  audioComponentDescription.componentSubType = 'Wotj';
  audioComponentDescription.componentManufacturer = 'InMo';
  audioComponentDescription.componentFlags = 0;
  audioComponentDescription.componentFlagsMask = 0;

  // Get the Wotja AUv3 unit
  AudioComponent audioComponent = AudioComponentFindNext(nil, &audioComponentDescription);
  if (audioComponent == nil) {
    sWotjaStatus = WotjaStatus_not_found;

    completion(WotjaStatus_not_found);

    return;
  }

  // Extract version string
  UInt32 versionNumber = 0;
  AudioComponentGetVersion(audioComponent, &versionNumber);
  sWotjaVersion = versionNumberToString(versionNumber);

  UInt32 minimumRequiredVersionNumber = versionNumberFromString(sWotjaMinimumVersionRequired);
  if (versionNumber < minimumRequiredVersionNumber) {
    completion(WotjaStatus_needs_update);
  } else {
    completion(WotjaStatus_found);
  }
}

// MARK: Wotja installation, update and miscellaneous

+(void) installWotja {
  [self sOpenURLInForeground:@"https://itunes.apple.com/app/id1344821101"];
}

+(void) updateWotja {
  [self sOpenURLInForeground:@"https://itunes.apple.com/app/id1344821101"];
}

+(void) showWotjaEULA {
  [self sOpenURLInForeground:@"https://wotja.com/legal/eula"];
}

+(void) showDeveloperFAQ {
  [self sOpenURLInForeground:@"https://wotja.com/help/#developer"];
}

+(void) showAboutWotja {
  [self sOpenURLInForeground:@"https://wotja.com/music/"];
}

+(void) showWotja {
  [self sOpenURLInForeground:@"wotja://api_show/"];
}

@end
